File: frontend/src/api/backup.ts
/**
 * Phase 3.5: backup status API client.
 *
 * Endpoints exposed by backend `app/api/backup.py` (mounted under
 * `/api/v1`). All requests go through `@/api/request` so the JWT
 * bearer token is injected automatically.
 */

import request from './request'

export interface BackupStatusLatest {
  id: number
  instance_id: number
  instance_name: string
  db_type_code: string
  host: string
  port: number
  backup_type: string | null
  source_type: string
  last_status: string
  last_success_at: string | null
  last_failure_at: string | null
  recovery_point_at: string | null
  age_minutes: number | null
  duration_seconds: number | null
  backup_size_mb: number | null
  message: string | null
  collected_at: string
  collector_run_id: number | null
  collector_run_item_id: number | null
}

export interface BackupStatusHistory {
  id: number
  instance_id: number
  backup_type: string | null
  source_type: string
  last_status: string
  last_success_at: string | null
  last_failure_at: string | null
  recovery_point_at: string | null
  age_minutes: number | null
  duration_seconds: number | null
  backup_size_mb: number | null
  message: string | null
  evidence: Record<string, unknown>
  collected_at: string
  created_at: string
}

export interface BackupCollectPayload {
  instance_ids: number[]
  db_type_code: string
  backup_type?: string | null
  sql_text: string
  timeout_seconds?: number
  max_rows?: number
}

export interface BackupCollectResponse {
  collector_run_id: number
  status: string
  awx_job_id: number | null
}

export interface BackupValidateSqlPayload {
  db_type_code: string
  sql_text: string
}

export interface BackupValidateSqlResponse {
  valid: boolean
  sql_hash: string
  message: string
  errors: string[]
}

export interface BackupLatestFilters {
  db_type_code?: string
  last_status?: string
  backup_type?: string
  keyword?: string
  limit?: number
}

export const backupApi = {
  listLatest(filters: BackupLatestFilters = {}): Promise<BackupStatusLatest[]> {
    return request.get('/v1/backup/status/latest', { params: filters })
  },
  listHistory(params: { instance_id?: number; limit?: number } = {}): Promise<BackupStatusHistory[]> {
    return request.get('/v1/backup/status/history', { params })
  },
  collect(payload: BackupCollectPayload): Promise<BackupCollectResponse> {
    return request.post('/v1/backup/status/collect', payload)
  },
  validateSql(payload: BackupValidateSqlPayload): Promise<BackupValidateSqlResponse> {
    return request.post('/v1/backup/status/validate-sql', payload)
  },
}

export default backupApi

File: backend/app/schemas/backup.py
"""Pydantic schemas for backup status endpoints (Phase 3.5)."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Optional

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Request: launch a backup status collect
# ---------------------------------------------------------------------------


class BackupCollectRequest(BaseModel):
    instance_ids: list[int] = Field(min_length=1, max_length=200)
    db_type_code: str = Field(min_length=1, max_length=50)
    backup_type: Optional[str] = Field(default=None, max_length=50)
    sql_text: str = Field(min_length=1)
    timeout_seconds: int = Field(default=30, ge=1, le=120)
    max_rows: int = Field(default=200, ge=1, le=1000)


class BackupCollectResponse(BaseModel):
    collector_run_id: int
    status: str
    awx_job_id: Optional[int] = None


# ---------------------------------------------------------------------------
# Latest backup status (joined with db_instance/db_type/server)
# ---------------------------------------------------------------------------


class BackupStatusLatestItem(BaseModel):
    id: int
    instance_id: int
    instance_name: str
    db_type_code: str
    host: str
    port: int
    backup_type: Optional[str] = None
    source_type: str
    last_status: str
    last_success_at: Optional[datetime] = None
    last_failure_at: Optional[datetime] = None
    recovery_point_at: Optional[datetime] = None
    age_minutes: Optional[int] = None
    duration_seconds: Optional[int] = None
    backup_size_mb: Optional[float] = None
    message: Optional[str] = None
    collected_at: datetime
    collector_run_id: Optional[int] = None
    collector_run_item_id: Optional[int] = None


# ---------------------------------------------------------------------------
# History (raw snapshot rows)
# ---------------------------------------------------------------------------


class BackupStatusHistoryItem(BaseModel):
    id: int
    instance_id: int
    backup_type: Optional[str] = None
    source_type: str
    last_status: str
    last_success_at: Optional[datetime] = None
    last_failure_at: Optional[datetime] = None
    recovery_point_at: Optional[datetime] = None
    age_minutes: Optional[int] = None
    duration_seconds: Optional[int] = None
    backup_size_mb: Optional[float] = None
    message: Optional[str] = None
    evidence: dict[str, Any] = Field(default_factory=dict)
    collected_at: datetime
    created_at: datetime


# ---------------------------------------------------------------------------
# Validate-SQL (defensive; same gate as inspection)
# ---------------------------------------------------------------------------


class ValidateBackupSqlRequest(BaseModel):
    db_type_code: str = Field(min_length=1, max_length=50)
    sql_text: str = Field(min_length=1)


class ValidateBackupSqlResponse(BaseModel):
    valid: bool
    sql_hash: str
    message: str
    errors: list[str] = Field(default_factory=list)
